### ISSUE TITLE
No published DTD or XSD for testng-results.xml makes it difficult to parse

### ISSUE THREAD

**Issue by ericdriggs** [testng-team/testng#2371]
### TestNG Version
N/A

### Expected behavior
published DTD for testng-results.xml to enable parsing of results so software can determine whether a given test has passed or failed

### Actual behavior
Only DTD published is for test-suite, which contains no attributes relevant to whether tests passed or failed

### Is the issue reproductible on runner?
N/A

### Test case sample
N/A

### Notes: 

XSD is preferred - most code generation tools require XSD so a DTD would need to be converted to XSD (As is the case with the current test suite DTD)

**Comment by ericdriggs** [comment]
Using the sample xml from: https://www.dev2qa.com/testng-report-example/
and the xsd generator from: https://xml.mherman.org/

I was able to generate the following, but I haven't verified for completeness:

testng-results.xsd
`<?xml version="1.0" encoding="UTF-8"?>
<xs:schema xmlns:xs="http://www.w3.org/2001/XMLSchema" elementFormDefault="qualified">
  <xs:element name="testng-results">
    <xs:complexType>
      <xs:sequence>
        <xs:element ref="reporter-output"/>
        <xs:element maxOccurs="unbounded" ref="suite"/>
      </xs:sequence>
      <xs:attribute name="failed" use="required" type="xs:integer"/>
      <xs:attribute name="ignored" use="required" type="xs:integer"/>
      <xs:attribute name="passed" use="required" type="xs:integer"/>
      <xs:attribute name="skipped" use="required" type="xs:integer"/>
      <xs:attribute name="total" use="required" type="xs:integer"/>
    </xs:complexType>
  </xs:element>
  <xs:element name="suite">
    <xs:complexType>
      <xs:sequence>
        <xs:element ref="groups"/>
        <xs:element maxOccurs="unbounded" ref="test"/>
      </xs:sequence>
      <xs:attribute name="duration-ms" use="required" type="xs:integer"/>
      <xs:attribute name="finished-at" use="required" type="xs:NMTOKEN"/>
      <xs:attribute name="name" use="required" type="xs:NCName"/>
      <xs:attribute name="started-at" use="required" type="xs:NMTOKEN"/>
    </xs:complexType>
  </xs:element>
  <xs:element name="groups">
    <xs:complexType>
      <xs:sequence>
        <xs:element minOccurs="0" maxOccurs="unbounded" ref="group"/>
      </xs:sequence>
    </xs:complexType>
  </xs:element>
  <xs:element name="group">
    <xs:complexType>
      <xs:sequence>
        <xs:element maxOccurs="unbounded" ref="method"/>
      </xs:sequence>
      <xs:attribute name="name" use="required" type="xs:NCName"/>
    </xs:complexType>
  </xs:element>
  <xs:element name="method">
    <xs:complexType>
      <xs:attribute name="class" use="required"/>
      <xs:attribute name="name" use="required" type="xs:NCName"/>
      <xs:attribute name="signature" use="required"/>
    </xs:complexType>
  </xs:element>
  <xs:element name="test">
    <xs:complexType>
      <xs:sequence>
        <xs:element maxOccurs="unbounded" ref="class"/>
      </xs:sequence>
      <xs:attribute name="duration-ms" use="required" type="xs:integer"/>
      <xs:attribute name="finished-at" use="required" type="xs:NMTOKEN"/>
      <xs:attribute name="name" use="required" type="xs:NCName"/>
      <xs:attribute name="started-at" use="required" type="xs:NMTOKEN"/>
    </xs:complexType>
  </xs:element>
  <xs:element name="class">
    <xs:complexType>
      <xs:sequence>
        <xs:element maxOccurs="unbounded" ref="test-method"/>
      </xs:sequence>
      <xs:attribute name="name" use="required"/>
    </xs:complexType>
  </xs:element>
  <xs:element name="test-method">
    <xs:complexType>
      <xs:sequence>
        <xs:choice minOccurs="0">
          <xs:element ref="exception"/>
          <xs:element ref="params"/>
        </xs:choice>
        <xs:element ref="reporter-output"/>
      </xs:sequence>
      <xs:attribute name="data-provider" type="xs:NCName"/>
      <xs:attribute name="duration-ms" use="required" type="xs:integer"/>
      <xs:attribute name="finished-at" use="required" type="xs:NMTOKEN"/>
      <xs:attribute name="name" use="required" type="xs:NCName"/>
      <xs:attribute name="signature" use="required"/>
      <xs:attribute name="started-at" use="required" type="xs:NMTOKEN"/>
      <xs:attribute name="status" use="required" type="xs:NCName"/>
    </xs:complexType>
  </xs:element>
  <xs:element name="exception">
    <xs:complexType>
      <xs:sequence>
        <xs:element minOccurs="0" ref="message"/>
        <xs:element ref="full-stacktrace"/>
      </xs:sequence>
      <xs:attribute name="class" use="required" type="xs:NCName"/>
    </xs:complexType>
  </xs:element>
  <xs:element name="message" type="xs:string"/>
  <xs:element name="full-stacktrace" type="xs:string"/>
  <xs:element name="params">
    <xs:complexType>
      <xs:sequence>
        <xs:element ref="param"/>
      </xs:sequence>
    </xs:complexType>
  </xs:element>
  <xs:element name="param">
    <xs:complexType>
      <xs:sequence>
        <xs:element ref="value"/>
      </xs:sequence>
      <xs:attribute name="index" use="required" type="xs:integer"/>
    </xs:complexType>
  </xs:element>
  <xs:element name="value" type="xs:string"/>
  <xs:element name="reporter-output">
    <xs:complexType>
      <xs:sequence>
        <xs:element minOccurs="0" maxOccurs="unbounded" ref="line"/>
      </xs:sequence>
    </xs:complexType>
  </xs:element>
  <xs:element name="line" type="xs:string"/>
</xs:schema>
`

**Comment by wengm** [comment]
Isn't this enough or could you please clarify what you need?
`<testng-results skipped="5" failed="1" ignored="24" total="74" passed="44">`
```
<test-method status="SKIP"
<test-method status="PASS"
<test-method status="PASS"
<test-method status="FAIL"
```
Do you need to show the test result in &lt;test&gt; level?

**Comment by ericdriggs** [comment]
Ideally, I'd like a canonical XSD or DTD published and available from testng's site so I and others can use code generation for parsing. What I did above was just a small stab at that, as evidenced by the NCNames instead of enumerations.

**Comment by wengm** [comment]
I agree it's better to have one 'official' DTD/XSD for testng-results.xml just like testng-1.0.dtd for suite XML file. 
@krmahadevan What do you think?

**Comment by krmahadevan** [comment]
@cbeust - Your thoughts ?

**Comment by juherr** [comment]
AI triage verdict: **still relevant**.

The current repository still has no canonical `testng-results.xsd` or `testng-results.dtd`, and there is no `GITHUB-2371` implementation/history marker. `testng-results.xml` is still produced by the XML reporters and is covered by current reporter tests, but those tests validate generated content rather than a published schema. The newer `testng-1.1.xsd` work and #3163 concern **suite XML**, not `testng-results.xml`, so they do not supersede this request. I found no open or merged PR implementing a results schema.

The historical reproducer was not executed; this issue is a missing published contract rather than a runtime reproducer. The current absence is directly established from the repository.

Next step: decide the compatibility/versioning contract for a canonical results schema, then derive an XSD from the current `XMLReporter` output and add validation fixtures covering success/failure/skip, parameters, exceptions, groups and reporter output before publishing it.
